### DISCUSSION TITLE
Unable to turn white in transparent without getting ugly edges

### DISCUSSION THREAD

**Discussion by digitalforgeangelo** [ImageMagick/ImageMagick#5146] [Help]
Hey guys, I am trying to remove the white from 5k+ pictures and turn it into transparent. Unfortunately, when I do this it leaves a ugly edge around the black parts in various gray tones.

I tried various methods including -trim and -fuzz but I am unable to get it working properly. All it should do is remove the white and leave everything else intract.

Can someone help me out please?

The original image:
https://i.stack.imgur.com/f456w.png

The image after using ImageMagick (added green so you can see the ugly edge better):
https://i.stack.imgur.com/9gSWC.png

The image after using ImageMagick (original)
https://i.stack.imgur.com/NDO0f.png

So basically what I want is to remove all white from the pictures and make those transparent. Anything else should NOT be affected.

The image was made in 8k resolution in Illustrator which I then exported to a single PNG image (transparent). Then I used Bramus' tile cutter script for Photoshop to cut my image into tiles (which forces a white background unfortunately).

So finally I wanted to try ImageMagick to remove the white but as you can read, I am unable to so far. 

I have to crank the fuzz so high for the arrows to remotely look a little bit properly, that ImageMagick will remove other stuff that it shouldn't as well.

Hope someone can help me out.

My current code:  
```
for /R "C:\Users\<Username>\Desktop\map" %I in (*.png) do magick "%I" -fuzz 69% -transparent #FFFFFF "%I"
```

**Comment by snibgo** [comment]
Look carefully at your image. It has black marks on a white background, with shades of gray at the boundary. When you make white pixels transparent, this doesn't change those gray pixels, the "ugly edge around the black parts in various gray tones."

What do you want to happen to those gray pixels?

Possibilities include:

1. Leave the gray pixels unchanged.

2. Make them fully transparent.

3. Make them black, with varying transparency.

4. Something else.

> So basically what I want is to remove all white from the pictures and make those transparent. Anything else should NOT be affected.

Is that what you want? You want the gray pixels unchanged?

**Reply by digitalforgeangelo** [reply]
Hi snibgo, thank you for your fast reply! I am unsure what would look the best, but I think the best option would be to make them black with varying transparency. I would like to get an outcome as close as possible to the Illustrator version that I made... but I understand that this is not possible as a PNG is not a Vector ;) 

What it should look like in the end (when on top of my other layer):
![image](https://user-images.githubusercontent.com/85583885/167883478-a6e34576-3521-4396-b7a4-a48a87176ffd.png)

**Comment by snibgo** [comment]
Here is a hacked solution. If it does what you want, it can be tidied up into a single command. If it doesn't do what you want, it can be hacked further.

Windows BAT script.
```
magick ^
  f456w.png ^
  -fuzz 10%% ^
  -transparent White ^
  +write x0.png ^
  -alpha extract ^
  +write x1.png ^
  -blur 0x0.5 ^
  mask.png

magick ^
  f456w.png ^
  mask.png ^
  -alpha off ^
  -compose CopyAlpha -composite ^
  out1.png

rem Make partially-transparent pixels black.
magick ^
  out1.png ^
  ( +clone -background Black -colorize 100 ) ^
  ( mask.png ^
    -fill Black -opaque White ^
    -fill White +opaque Black ^
  ) ^
  -alpha off ^
  -compose Over -composite ^
  mask.png ^
  -alpha off ^
  -compose CopyAlpha -composite ^
  out2.png

rem Test the edges.
magick ^
  out2.png ^
  -background #0f0 ^
  -layers Flatten ^
  f2.png
```
out2.png is:
![out2](https://user-images.githubusercontent.com/33812368/167890989-2570fc09-5218-4242-932d-2a35d8e9f38b.png)

**Reply by digitalforgeangelo** [reply]
Ooohhhh, you are a hero! I will try to get this working on my end and let you know the outcome! Thank you very, very much!!!

P.S. Do you think it would be possible to make the edges a little bit smoother or is that (near) impossible? they look a bit... rough? :D 

Anyways, thanks so much for your reply! really, really appreciated!

**Comment by snibgo** [comment]
> P.S. Do you think it would be possible to make the edges a little bit smoother or is that (near) impossible? they look a bit... rough? :D

Sure. Change "-blur 0x0.5" to "-blur 0x1.0". If that is too blurry, try "-blur 0x0.75" or some other number.

**Reply by digitalforgeangelo** [reply]
Seems to work prefectly @snibgo , really awesome! I just need to get this working now with this:

```
for /R "C:\Users\<Username>\Desktop\map\reserve_floors" %I in (*.png) do magick "%I" -fuzz 10% -transparent #FFFFFF "%I"
```

This recursively loops through all folders and does the magic for each image :D Trying to get them both in the same script, but keeps erroring hahaha. It is also very important that the output name is not output2 but the same as the input. Names should not change.

**Comment by snibgo** [comment]
If that is in a BAT script, you need to double the percent signs.

**Reply by digitalforgeangelo** [reply]
Yeah I tried that but it is not working when using multiple lines. When I put everything on a single line it is working but the file names get messed up lol.

```
@echo off

for /R "C:\Users\<Username>\Desktop\map\reserve_floors" %%I in (*.png) do ^

magick %%I -fuzz 10%% -transparent White +write x0.png -alpha extract +write x1.png -blur 0x0.5 mask.png ^
magick %%I mask.png -alpha off -compose CopyAlpha -composite temp.png ^
magick temp.png ( +clone -background Black -colorize 100 ) ( mask.png -fill Black -opaque White -fill White +opaque Black ) -alpha off -compose Over -composite mask.png -alpha off -compose CopyAlpha -composite %%I

pause
```

For readability its not on one single line obviously ;) 

The above gives me an error that "mask.png" does not exist. Anyways, this is approximately what I would like to achieve.
Loop through all images, convert them (with the same name!!!) like you did with the out2.png which looks near perfect and then delete the temporarily files so I don't have to do it manually.

I will keep trying to get it working but... don't think I will get it working haha.

**Comment by snibgo** [comment]
Oh dear.

Why do you have "@echo off"? That makes debugging more difficult.

Caret "^" at the end of a line means the command is continued on the next line. The BAT interpreter will join those lines together to make a complete command. The lines beginning with "magick" are separate commands, so should not have "^" between them.

All lines with "%%I" should be inside the "for" loop. "for ... in ... do" can be followed by a single command, or by any number of commands in parentheses "(...)". As you want multiple commands in the loop, you need parentheses.

Any magick command within the "for (...)" loop must have any parentheses escaped with a caret.

So, it should be more like this:
```
for /R "C:\Users\<Username>\Desktop\map\reserve_floors" %%I in (*.png) do (

  magick %%I -fuzz 10%% -transparent White +write x0.png -alpha extract +write x1.png -blur 0x0.5 mask.png

  magick %%I mask.png -alpha off -compose CopyAlpha -composite temp.png

  magick ^
    temp.png ^
    ^( +clone -background Black -colorize 100 ^) ^
    ^( mask.png -fill Black -opaque White -fill White +opaque Black ^) ^
    -alpha off -compose Over -composite mask.png -alpha off -compose CopyAlpha -composite %%I

)

pause
```
I have added blank lines and indentation, and chopped the final magick command into separate lines, for clarity.
